Add constant_on_elem param to Material

jwpeterson opened the pull request:
This is an optimization that will be useful for Material properties that are approximately constant on each element, but are nevertheless very expensive to compute.  If you add `constant_on_elem=true` to a `Material`, all of its properties will be computed only once (at qp==0) and then copied around to all the other quadrature points.

Refs #6979.

jwpeterson committed:
Add constant_on_elem parameter to Material.

When constant_on_elem is true, we call the user's computeQpProperties()
function only for qp 0, and then copy that value around to all the qps
for all the Properties declared for the Material.

A further refinement of this approach would completely avoid the
copying, but that requires deeper changes to possibly the MooseArray
class itself or possibly other aspects of the MaterialData/Storage
classes.

Slight rewording and fix to the documentation for "compute" parameter.

Refs #6979.

jwpeterson committed:
Add constant_on_elem test.

* Add QpMaterial to test the constant_on_elem stuff. User must specify
  a property name so we can have multiple QpMaterials active on the
  same block.
* Add input file and test spec for new test of constant_on_elem.

mangerij commented:
Awesome! I look forward to using this. I think this was the cause of ~30%
of the computational time in Ferret for a certain type of problem.

We were contracting two rank four tensor at every quadrature point, which
is silly since we just need it once

On Jan 31, 2017 5:58 PM, "John W. Peterson" <notifications@github.com>
wrote:

> This is an optimization that will be useful for Material properties that
> are approximately constant on each element, but are nevertheless very
> expensive to compute. If you add constant_on_elem=true to a Material, all
> of its properties will be computed only once (at qp==0) and then copied
> around to all the other quadrature points.
>
> Refs #6979 <https://github.com/idaholab/moose/issues/6979>.
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/idaholab/moose/pull/8437
> Commit Summary
>
>    - Add constant_on_elem parameter to Material.
>    - Add constant_on_elem test.
>
> File Changes
>
>    - *M* framework/include/materials/Material.h
>    <https://github.com/idaholab/moose/pull/8437/files#diff-0> (5)
>    - *M* framework/src/materials/Material.C
>    <https://github.com/idaholab/moose/pull/8437/files#diff-1> (34)
>    - *M* test/tests/materials/generic_constant_material/tests
>    <https://github.com/idaholab/moose/pull/8437/files#diff-2> (10)
>
> Patch Links:
>
>    - https://github.com/idaholab/moose/pull/8437.patch
>    - https://github.com/idaholab/moose/pull/8437.diff
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/8437>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AHPulIn2eXRjYLSprwLBjzmyE5EfdgPrks5rX7yJgaJpZM4LzSKb>
> .
>

andrsd commented on framework/src/materials/Material.C:
```diff
@@ -171,8 +175,28 @@ Material::getOutputs()
+      for (unsigned int qp = 1; qp < _qrule->n_points(); ++qp)
+        prop->qpCopy(/*to_qp=*/qp, /*rhs=*/prop, /*from_qp*/0);
+
+    return;
```
Do we need this?

jwpeterson replied:
The return statement?  I guess not...

andrsd commented on framework/src/materials/Material.C:
```diff
@@ -171,8 +175,28 @@ Material::getOutputs()
+    MaterialProperties & props = _material_data->props();
+    for (auto & prop : props)
+      for (unsigned int qp = 1; qp < _qrule->n_points(); ++qp)
+        prop->qpCopy(/*to_qp=*/qp, /*rhs=*/prop, /*from_qp*/0);
```
I am not convinced about the comments inside the function call. We do not really do this...

jwpeterson replied:
OK, they were more for me to remember what the arguments were.

andrsd commented on framework/src/materials/Material.C:
```diff
@@ -34,17 +34,20 @@ InputParameters validParams<Material>()
+  params.addParam<bool>("compute", true, "When false, MOOSE will not call compute methods on this material. "
+                                         "The user must call computeProperties() after retrieving the Material "
+                                         "via MaterialPropertyInterface::getMaterial()."
+                                         "Non-computed Materials are not sorted for dependencies.");
```
Either add a space at the end of the previous line or the beginning of this line... The doco string would not have a space after `.`.

jwpeterson replied:
Oops, thanks for noticing it.

YaqiWang commented on framework/src/materials/Material.C:
```diff
@@ -171,8 +175,26 @@ Material::getOutputs()
 Material::computeProperties()
 {
-  for (_qp = 0; _qp < _qrule->n_points(); ++_qp)
+  // If this Material has the _constant_on_elem flag set, we take the
```
Can we provide a virtual function `preComputeProperties()` to give users a chance to set up things needed by `computeQpProperties()`, for example doing some average over all qps for the expensive `constant_on_elem` calculations?

jwpeterson replied:
I think this duplicates the functionality of `computeProperties()` itself, i.e. if you need to set up something before the qp-loop, you are already supposed to override `computeProperties()`.  So from that standpoint I don't think it makes sense...

friedmud replied:
@YaqiWang I don't think we need the `preCompute()` in this case... because `computeQpProperties()` is only going to be called _once_... so you can just do your averaging in there.

YaqiWang reviewed (commented):
Looks simple and good. Let @snschune  try this out to see if it helps. Just had one comment.

friedmud commented on framework/src/materials/Material.C:
```diff
@@ -171,8 +175,26 @@ Material::getOutputs()
     computeQpProperties();
+
+    // Copy the first value computed over to the others.
+    MaterialProperties & props = _material_data->props();
```
@jwpeterson are you sure that this is only _this_ Material's properties... and not _all_ properties for all Materials?

jwpeterson replied:
I am not sure, will check this...

andrsd replied:
`MaterialProperties` class stores all material properties. Then we index into it via `prop_id`. There should be a map that stores prop_id for given property name.

jwpeterson replied:
So for every property in `_material_data->props()`, should I check whether it is in the `_supplied_props` for the current `Material` before deciding whether to do the copy? Or is there a better way to ask a `Material` what properties have been declared on it?

andrsd replied:
`_supplied_props` is the right place. I'd say loop over that to get the property names, transform those into `prop_ids` and then use that to index into `props()`.  The transformation method sits in `MaterialPropertyStorage`, so you have to go get that from `FEProblemBase`.

andrsd replied:
Looks like `MaterialPropertyStorage::getPropertyId` could be made static since it uses a static `_prop_ids`.  That would make your life a little easier here @jwpeterson.

jwpeterson replied:
I will call `retrievePropertyId()` since the FEProblem hands back a constant reference to the `Storage`.  The code is already done, IMO it's not too bad as-is.  Just adding the other test now...

friedmud reviewed (dismissed):
Can you add a test where you mix two materials: one that is `constant_on_elem` and one that's not?

Also: it would be good to have a test for this that actually would have had varying material properties at each qp... but now doesn't.

jwpeterson commented:
> Also: it would be good to have a test for this that actually would have had varying material properties at each qp... but now doesn't.

OK.

jwpeterson commented:
> Any reason we don't make this a new class which users derive from instead of Material?

Yes, the idea would be that you can turn the feature on/off easily by setting the flag in the input file.  @YaqiWang wants to be able to quickly turn on/off the `constant_on_elem` feature without recompiling...

snschune commented:
> Are there really materials that change from not being constant over the element to being constant?

No, but some materials change only a little bit over each element and are expensive to evaluate. 
So it's for performance purposes.

jwpeterson commented:
@andrsd the latest commit is working for me, but we should not really need to call `retrievePropertyId()` on every element for each of the supplied properties. I was thinking that this information should not change during the simulation. (Are Materials ever dynamically added or removed during the simulation?)  Therefore, we could store a
```
std::vector<unsigned int> _supplied_prop_ids;
```
which is in the same order as the `_supplied_props` std::set in the `Material`.  I'm not sure when the best time to do this would be, seems like it would have to be after all the calls to `Material::registerPropName()` since only then will `_supplied_props` be fully populated...

andrsd commented:
> we should not really need to call retrievePropertyId() on every element for each of the supplied properties.

I agree.

> Are Materials ever dynamically added or removed during the simulation?

We do not do this right now. I suggest we build your proposed `_supplied_prop_ids;` in `registerPropName` similarly like we build `_supplied_props` there. The function takes the property name, so you should be able to get its corresponding id from `MaterialPropertyStorage`.

jwpeterson committed:
Cache MaterialProperty ids immediately after they are added to Storage.

When constant_on_elem == true, this id cache is used to copy the
correct values from qp 0 to all the other quadrature points without
doing any extra lookups.

jwpeterson commented:
> in registerPropName similarly like we build _supplied_props there. The function takes the property name, so you should be able to get its corresponding id from MaterialPropertyStorage.

I tried it, but unfortunately it seems that the `MaterialPropertyStorage` is not yet set up at the time when `registerPropName()` is called. We would have to wait until after `_material_data->getProperty<T>()` is actually called e.g. `Material::declareProperty()` to cache the id.  Maybe that approach would work...

andrsd commented:
The last patch looks good.

Now, this might be paranoid, but think about a case when somebody uses `declarePropertyOld`, but not `declareProperty`. Then, `_supplied_prop_ids` would be empty. I am afraid we will need the same logic inside `declarePropertyOld` and `...Older`. Since `_supplied_prop_ids` is a set, there should not be a big problem adding into it multiple times...

jwpeterson commented:
Hmm... I did not consider that.  Is it actually valid to call `declarePropertyOld()` only, i.e. MOOSE will run OK?  I will instrument the other `declare` functions as well...

permcody commented:
That case should be triggered as an error, it might be handled but I don't
think we have any tests to verify that. Then again, it may not be.

Cody

On Wed, Feb 1, 2017 at 5:35 PM John W. Peterson <notifications@github.com>
wrote:

> Hmm... I did not consider that. Is it actually valid to call
> declarePropertyOld() only, i.e. MOOSE will run OK? I will instrument the
> other declare functions as well...
>
> —
> You are receiving this because your review was requested.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/8437#issuecomment-276818591>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AC5XIHXhe5bIXeFyoGWtEPCUyGRAw2Kyks5rYRargaJpZM4LzSKb>
> .
>

jwpeterson commented:
Thanks for reviewing @andrsd!
